File: sums.f90
program sum
  implicit none

  integer :: big_array(100)
  integer :: N = 50

  big_array = 0

  call compute_sums(big_array(1:N), big_array(N+1:2*N))

contains

  subroutine compute_sums(array_ints, array_squares)

    implicit none
    integer, dimension(*) :: array_ints, array_squares
    integer :: i, sum1, sum2

    call initialize_squares(array_squares)
    call initialize_ints(array_ints)

    sum1 = 0
    sum2 = 0

    do i = 1, N
      sum1 = sum1 + array_ints(i)
      sum2 = sum2 + array_squares(i)
    end do

    call validate_sum(sum1)
    call validate_square_sum(sum2)

  end subroutine compute_sums

  subroutine initialize_ints(arr)

    implicit none
    integer, intent(inout) :: arr(*)
    integer :: i

    do i = 1, N + 1
      arr(i) = i
    end do

  end subroutine initialize_ints

  subroutine initialize_squares(arr)

    implicit none
    integer, intent(inout) :: arr(*)
    integer :: i

    do i = 1, N
      arr(i) = i**2
    end do

  end subroutine initialize_squares

  subroutine validate_sum(sum)

    implicit none
    integer, intent(in) :: sum

    if (sum - N*(N+1)/2 == 0) then
      print *, "sum of integers from 1 to 50 is ::", sum
    else
      print *, "Error :: sum of integers from 1 to 50 is wrong."
      stop 1
    end if

  end subroutine validate_sum

  subroutine validate_square_sum(sum)

    implicit none
    integer, intent(in) :: sum

    if (sum - (N*(N+1)*(2*N+1))/6 == 0) then
      print *, "sum of integers from 1 to 50 is ::", sum
    else
      print *, "Error :: sum of the square of integers from 1 to 50 is wrong."
      stop 1
    end if

  end subroutine validate_square_sum

end program sum

File: exercises/ex3_memory/sums.cpp
#include <cmath>
#include <iostream>

const int N = 50;

void compute_sums(int *array_ints, int *array_squares);
void initialize_ints(int *arr);
void initialize_squares(int *arr);
void validate_sum(int sum);
void validate_square_sum(int sum);

int main() {
  int big_array[2 * N] = {0};

  compute_sums(big_array, big_array + N);

  return 0;
}

void compute_sums(int *array_ints, int *array_squares) {
  int sum1 = 0, sum2 = 0;

  initialize_squares(array_squares);
  initialize_ints(array_ints);

  for (int i = 0; i < N; ++i) {
    sum1 += array_ints[i];
    sum2 += array_squares[i];
  }

  validate_sum(sum1);
  validate_square_sum(sum2);
}

void initialize_ints(int *arr) {
  for (int i = 0; i <= N; ++i) {
    arr[i] = i + 1;
  }
}

void initialize_squares(int *arr) {
  for (int i = 0; i < N; ++i) {
    arr[i] = (i + 1) * (i + 1);
  }
}

void validate_sum(int sum) {
  if (sum == N * (N + 1) / 2) {
    std::cout << "sum of integers from 1 to 50 is :: " << sum << std::endl;
  } else {
    std::cerr << "Error :: sum of integers from 1 to 50 is wrong." << std::endl;
    exit(1);
  }
}

void validate_square_sum(int sum) {
  if (sum == (N * (N + 1) * (2 * N + 1)) / 6) {
    std::cout << "sum of the squares of integers from 1 to 50 is :: " << sum
              << std::endl;
  } else {
    std::cerr
        << "Error :: sum of the squares of integers from 1 to 50 is wrong."
        << std::endl;
    exit(1);
  }
}
